### PR TITLE
Update testPrefixScan to work with different warp sizes

### PR DESCRIPTION
**Pull request by fwyzard**
#### PR description:

Update the alpaka-based testPrefixScan to work with different warp sizes.

#### PR validation:

Unit test now passes on CPU, CUDA and ROCm back-ends.

### CODE REVIEW

File: HeterogeneousCore/AlpakaInterface/test/alpaka/testPrefixScan.dev.cc [hunk 1]
```diff
@@ -34,7 +34,17 @@ template <typename T>
...
+#if defined(__CUDA_ARCH__)
+    // CUDA always has a warp size of 32
     auto& ws = alpaka::declareSharedVar<T[32], __COUNTER__>(acc);
+#elif defined(__HIP_DEVICE_COMPILE__)
+    // HIP/ROCm defines warpSize as a constant expression with value 32 or 64 depending on the target device
+    auto& ws = alpaka::declareSharedVar<T[warpSize], __COUNTER__>(acc);
+#else
+    // CPU back-ends always have a warp size of 1
+    auto& ws = alpaka::declareSharedVar<T[1], __COUNTER__>(acc);
+#endif
```
Comment (makortel) [inline HeterogeneousCore/AlpakaInterface/test/alpaka/testPrefixScan.dev.cc:47]: Should we think of an abstraction for compile-time defined warp size?

I suppose alternatively these could use the dynamic shared memory.

File: HeterogeneousCore/AlpakaInterface/test/alpaka/testPrefixScan.dev.cc [hunk 2]
```diff
@@ -152,21 +163,26 @@ int main() {
...
-      const auto threadsPerBlockOrElementsPerThread = 32;
+      const auto threadsPerBlockOrElementsPerThread = warpSize;
       const auto blocksPerGrid = 1;
       const auto workDivWarp = make_workdiv<Acc1D>(blocksPerGrid, threadsPerBlockOrElementsPerThread);
 
-      alpaka::enqueue(queue, alpaka::createTaskKernel<Acc1D>(workDivWarp, testWarpPrefixScan<int>(), 32));
-      alpaka::enqueue(queue, alpaka::createTaskKernel<Acc1D>(workDivWarp, testWarpPrefixScan<int>(), 16));
-      alpaka::enqueue(queue, alpaka::createTaskKernel<Acc1D>(workDivWarp, testWarpPrefixScan<int>(), 5));
+      if (warpSize >= 64)
+        alpaka::enqueue(queue, alpaka::createTaskKernel<Acc1D>(workDivWarp, testWarpPrefixScan<int>(), 64));
```
Comment (makortel) [inline HeterogeneousCore/AlpakaInterface/test/alpaka/testPrefixScan.dev.cc:171]: While not about the fix itself, would it be worth of moving this test to `alpaka::exec`?

File: HeterogeneousCore/AlpakaInterface/test/alpaka/testPrefixScan.dev.cc (hunk 1 above)
Comment (fwyzard) [reply HeterogeneousCore/AlpakaInterface/test/alpaka/testPrefixScan.dev.cc:47]: Yes, it's something we are working on.
The main problem is that SYCL allows for different warp sizes, and the actual value is only known at _link_ time (when device compilation happens), not at compile time. 
We have a partially working solution, but it fails in some corner cases, and are discussing with Intel how to fix them. 
Once that is solved, we plan to implement a compile-time constant for the warp size in Alpaka.

**Comment by fwyzard** [comment]
+1

**Comment by mandrenguyen** [comment]
+1

**Pull request merged**